hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/ImmutableSegment.java

```diff
@@ -43,4 +48,127 @@ public KeyValueScanner getKeyValueScanner() {
     return new CollectionBackedScanner(getCellSet(), getComparator());
   }
 
+  public boolean isFlat() {
+    return isFlat;
+  }
+
+  protected ImmutableSegment(Segment segment) {
+    super(segment);
+    isFlat = false;
+  }
+
+  // C-tor by flattening other (old) segment
+  protected ImmutableSegment(ImmutableSegment oldSegment, MemStoreCompactorIterator iterator,
+      MemStoreLAB memStoreLAB, int numOfCells, long constantCellSizeOverhead) {
+
+    super(null,oldSegment.getComparator(),memStoreLAB,
+        CompactingMemStore.DEEP_OVERHEAD_PER_PIPELINE_FLAT_ARRAY_ITEM, constantCellSizeOverhead);
+    CellSet cs = this.createArrayBasedCellSet(numOfCells, iterator, false); // build the CellSet
+    this.setCellSet(cs);      // update the CellSet of the new Segment
+    isFlat = true;
+  }
+
+  // C-tor by compaction to CellArrayMap or CellChunkMap
+  protected ImmutableSegment(
+      final Configuration conf, CellComparator comparator, MemStoreCompactorIterator iterator,
+      MemStoreLAB memStoreLAB, int numOfCells, long constantCellSizeOverhead, boolean array) {
+
+    super(null, comparator, memStoreLAB,
+        CompactingMemStore.DEEP_OVERHEAD_PER_PIPELINE_FLAT_ARRAY_ITEM, constantCellSizeOverhead);
+    CellSet cs = null; // build the CellSet Cell array or Byte array based
+    if (array) {
+      cs = this.createArrayBasedCellSet(numOfCells, iterator, true);
+    } else {
+      cs = this.createChunkBasedCellSet(numOfCells, iterator, conf);
+    }
+    this.setCellSet(cs);  // update the CellSet of the new Segment
+    isFlat = true;
+  }
+
+  private CellSet createArrayBasedCellSet(
+      int numOfCells, MemStoreCompactorIterator iterator, boolean allocateFromMSLAB) {
+
+    Cell[] cells = new Cell[numOfCells];   // build the Cell Array
+    int i = 0;
+    while (iterator.hasNext()) {
+      Cell c = iterator.next();
+      if (allocateFromMSLAB) {
+        // The scanner behind the iterator is doing all the elimination logic
+        // now we just copy it to the new segment (also MSLAB copy)
+        KeyValue kv = KeyValueUtil.ensureKeyValue(c);
+        Cell newKV = maybeCloneWithAllocator(kv);
+        cells[i++] = newKV;
+        // flattening = false, compaction case, counting both Heap and MetaData size
+        updateMetaInfo(c,true,!allocateFromMSLAB);
+      } else {
+        cells[i++] = c;
+        // flattening = true, flattening case, counting only MetaData size
+        updateMetaInfo(c,true,!allocateFromMSLAB);
+      }
+
+    }
+    // build the immutable CellSet
+    CellArrayMap cam = new CellArrayMap(getComparator(),cells,0,i,false);
+    return new CellSet(cam);
+  }
+
+  private CellSet createChunkBasedCellSet(
+      int numOfCells, MemStoreCompactorIterator iterator, final Configuration conf) {
+
+    int chunkSize = conf.getInt(HeapMemStoreLAB.CHUNK_SIZE_KEY, HeapMemStoreLAB.CHUNK_SIZE_DEFAULT);
```

Comment:
this can be a static attribute that is set at construction time, only computed once